Sources/ShopifyCheckoutKit/AnalyticsEvents.swift

```diff
@@ -0,0 +1,3579 @@
+            cart: cart ?? self.cart,
+            checkout: checkout ?? self.checkout,
+            customer: customer ?? self.customer,
+            productVariants: productVariants ?? self.productVariants
+        )
+    }
+
+    func jsonData() throws -> Data {
+        return try newJSONEncoder().encode(self)
+    }
+
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// A customer represents a customer account with the shop. Customer accounts
+/// store contact information for the customer, saving logged-in customers the
+/// trouble of having to provide it at every checkout.
+// MARK: - Customer
+struct Customer: Codable {
+    /// The customer’s email address.
+    let email: String?
+    /// The customer’s first name.
+    let firstName: String?
+    /// The ID of the customer.
+    let id: String?
+    /// The customer’s last name.
+    let lastName: String?
+    /// The total number of orders that the customer has placed.
+    let ordersCount: Double?
+    /// The customer’s phone number.
+    let phone: String?
+}
+
+// MARK: Customer convenience initializers and mutators
+
+extension Customer {
+    init(data: Data) throws {
+        self = try newJSONDecoder().decode(Customer.self, from: data)
+    }
+
+    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
+        guard let data = json.data(using: encoding) else {
+            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
+        }
+        try self.init(data: data)
+    }
+
+    init(fromURL url: URL) throws {
+        try self.init(data: try Data(contentsOf: url))
+    }
+
+    func with(
+        email: String?? = nil,
+        firstName: String?? = nil,
+        id: String?? = nil,
+        lastName: String?? = nil,
+        ordersCount: Double?? = nil,
+        phone: String?? = nil
+    ) -> Customer {
+        return Customer(
+            email: email ?? self.email,
+            firstName: firstName ?? self.firstName,
+            id: id ?? self.id,
+            lastName: lastName ?? self.lastName,
+            ordersCount: ordersCount ?? self.ordersCount,
+            phone: phone ?? self.phone
+        )
+    }
+
+    func jsonData() throws -> Data {
+        return try newJSONEncoder().encode(self)
+    }
+
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// A value given to a customer when a discount is applied to an order. The
+/// application of a discount with this value gives the customer the specified
+/// percentage off a specified item.
+// MARK: - PricingPercentageValue
+struct PricingPercentageValue: Codable {
+    /// The percentage value of the object.
+    let percentage: Double?
+}
+
+// MARK: PricingPercentageValue convenience initializers and mutators
+
+extension PricingPercentageValue {
```

Comment:
Have we also covered DOM events?

Reply:
Are they already included in the Web Pixels Manager?

Reply:
Not yet, but think they will be very soon

Reply:
sure, we can come back to it